subsys/settings/src/settings_nvs.c

```diff
@@ -0,0 +1,238 @@
+/*
+ * Copyright (c) 2019 Laczen
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <string.h>
+
+#include "settings/settings.h"
+#include "settings/settings_nvs.h"
+#include "settings_priv.h"
+
+#include <logging/log.h>
+LOG_MODULE_DECLARE(settings, CONFIG_SETTINGS_LOG_LEVEL);
+
+struct settings_nvs_read_fn_arg {
+	struct nvs_fs *fs;
+	u16_t id;
+};
+
+static int settings_nvs_load(struct settings_store *cs, const char *subtree);
+static int settings_nvs_save(struct settings_store *cs, const char *name,
+			     const char *value, size_t val_len);
+
+static struct settings_store_itf settings_nvs_itf = {
+	.csi_load = settings_nvs_load,
+	.csi_save = settings_nvs_save,
+};
+
+static ssize_t settings_nvs_read_fn(void *back_end, void *data, size_t len)
+{
+	struct settings_nvs_read_fn_arg *rd_fn_arg;
+
+	rd_fn_arg = (struct settings_nvs_read_fn_arg *)back_end;
+
+	return nvs_read(rd_fn_arg->fs, rd_fn_arg->id, data, len);
+}
+
+int settings_nvs_src(struct settings_nvs *cf)
+{
+	cf->cf_store.cs_itf = &settings_nvs_itf;
+	settings_src_register(&cf->cf_store);
+
+	return 0;
+}
+
+int settings_nvs_dst(struct settings_nvs *cf)
+{
+	cf->cf_store.cs_itf = &settings_nvs_itf;
+	settings_dst_register(&cf->cf_store);
+
+	return 0;
+}
+
+static int settings_nvs_load(struct settings_store *cs, const char *subtree)
+{
+	struct settings_nvs *cf = (struct settings_nvs *)cs;
+	struct settings_nvs_read_fn_arg read_fn_arg;
+	struct settings_handler *ch;
+	char name[SETTINGS_MAX_NAME_LEN + SETTINGS_EXTRA_LEN + 1];
+	char buf;
+	const char *name_argv;
+	ssize_t rc1, rc2;
+	u16_t name_id = NVS_NAMECNT_ID;
+
+	/* In nvs the settings data is stored as:
+	 *	a. name at id starting from NVS_NAMECNT_ID + 1
+	 *	b. value at id of name + NVS_NAME_ID_OFFSET
+	 * Deleted records will not be found, only the last record will be
+	 * read. The entry ad NVS_NAMECNT_ID is used to store the largest
+	 * name id in use.
+	 */
+
+	name_id = cf->last_name_id + 1;
+
+	while (1) {
+
+		name_id--;
+		if (name_id == NVS_NAMECNT_ID) {
+			break;
+		}
+		rc1 = nvs_read(&cf->cf_nvs, name_id, &name, sizeof(name));
```

Comment:
I understand that saving the setting's name in one entry and the value in another (as opposed to saving everything in one entry) is done to optimize the usage of the flash, since once the application is compiled the name of the settings is basically static and it is only their value that changes. Thus, every time a setting's value is updated we will avoid rewriting the setting's name.

If my understanding is correct, it would be helpful to have a small comment here to remind why we are saving two entries instead of one.

Reply:
Yes, that's correct.

Added a short description to explain it.